### PR TITLE
RandomChoice[] and RandomSample[]

### PR DESCRIPTION
**Pull request by poke1024**
Adds RandomChoice[] and RandomSample[]. These two need numpy installed.

I rewired RandomInteger[] et al. to numpy (if it's installed), as there should be one consistent random number provider for the system that attaches to one random generator state, which can be set, reset, etc.. Starting to have different random generator states for e.g. RandomInteger[] and RandomChoice[] (and ImageRandom[] and others in the future for that matter) would surely spell disaster. So if numpy is available, everything is based on the numpy generator now.

If numpy is not installed (e.g. for PyPy) the implementation falls back and is equivalent in its results to the old implementation. Then, however, RandomChoice[] and RandomSample[] are not available. They could be added through adding a python only implementation in NoNumPyRandomEnv.randchoice().

I adjusted setup.py to pull the PyPy-numpy package for PyPy (https://bitbucket.org/pypy/numpy). The install works and downloads the package, but this then says that it needs PyPy4 to run, which is currently not supported by TravisCI. So I commented it out for now. It might be an option for the future though.

### CODE REVIEW

File: mathics/builtin/randomnumbers.py [hunk 1]
```diff
@@ -33,17 +65,60 @@ def get_random_state():
...
+            leaves = [new_element(x) for x in a]
+        else:
+            leaves = [_from_array(e, new_element, d) for e in a]
+        return Expression('List', *leaves)
+
+    def _stack_array(a):
+        if not isinstance(a[0], list):
+            return list(chain(a))
+        else:
+            return [_stack_array([x[i] for x in a]) for i in range(len(a[0]))]
```
Comment (sn6uv) [inline mathics/builtin/randomnumbers.py]: I wonder if these functions might be useful for other things that could use numpy, e.g. linear algebra routines. There's a short discussion on the mailing list see https://groups.google.com/forum/#!topic/mathics-devel/6btOa-5q87E but I think the best way to proceed would be a separate module e.g. `mathics/builtin/numpy_utils.py`. What do you think?

Comment (poke1024) [inline mathics/builtin/randomnumbers.py]: Sounds like a good idea. I'll move them.

**Comment by sn6uv** [comment]
In regards to the TravisCI config. Afaik pypy3 refers to the version of pypy supporting python 3. I think it would be better to try to install numpy under the standard version of pypy (and mark that as an expected failure).

**Comment by poke1024** [comment]
The Travis config now uses a new PYPY_NUMPY flag, and has an expected fail for pypy + PYPY_NUMPY.

File: mathics/builtin/randomnumbers.py [hunk 2]
```diff
@@ -13,17 +13,60 @@
...
+    _numpy = False
+    import random
+
+if _numpy:
+    import time
+    import os
+
+    # mathics.builtin.__init__.py module scanning logic gets confused
+    # if we assign numpy.random.get_state to a variable here. so we
+    # use defs to safely wrap the offending objects.
```
Comment (bnjones) [inline mathics/builtin/randomnumbers.py]: Can we fix this in the builtin loader, rather than working around it like this?

**Pull request merged**